### ISSUE TITLE
Slow for large repos

### ISSUE THREAD

**Issue by asl** [gitbutlerapp/gitbutler#2916]
Steps to reproduce: 
 - Grab a clone https://github.com/llvm/llvm-project
 - Importing as a new project takes several minutes at least
 - Then after selecting the main branch something takes minutes as well

**Comment by andrew-rosca** [comment]
Because many operations (e.g. creating a new virtual branch) take a very long time, and all controls remain enabled while an operation is in progress, the experience is very confusing with large repos.

For example, starting a commit takes several seconds. While the operation is in progress, there is no indication in the UI that something is happening in the background. This may cause you to click the "Start Commit" button multiple times, which results in the same operation being repeated and possibly errors.

**Comment by Byron** [comment]
That's very good to know, thanks so much for sharing!

Joining in @PavelLaptev, I thought I'd share I find this very interesting to solve from a UX perspective - buttons are surprisingly hard after all. If the operation is quick, the button might feel 'nervous' or flickery, so maybe it could enter a 'waiting' state only after a delay. Somehow I thought that if the user would click again within this delay, it could enter the 'waiting' state immediately to not let the user think nothing happen on click (somehow I feel UIs should always respond to a click, just to signal they received it and thus are working).

**Comment by PavelLaptev** [comment]
@Byron yep, that's a good idea. Let me check it first

**Comment by PavelLaptev** [comment]
@Byron I like the idea about timeout for "start commit" thing, in this case it wouldn't be "flickering".

In case you want to work, on this. The button has two props:
- `disabled` will only disable the button
- `loading` will disable the button and change or add the `spinner` icon

**Comment by PavelLaptev** [comment]
Made some edits to the project setup to prevent users from clicking on buttons while waiting: https://github.com/gitbutlerapp/gitbutler/pull/4090

**Comment by krlvi** [comment]
on the performance side of things, i have a pr that speeds things quite a bit https://github.com/gitbutlerapp/gitbutler/pull/4091

**Comment by Byron** [comment]
> @Byron while we can add more loading states, the core issue is more fundamental. For large projects, GitButler is not usable with such load times.

I agree, ideally any kind of loading time should be an edge-case (as GB is always fast enough to be instant). From a view of plain correctness, I think we should be able to assure that quick-clicks don't trigger a cascade of similar operations *or* that these operations are idempotent.

There is a part in me who thinks that if there was an abstraction for this, any 'action' button could be made to

* prevent interaction while the action is running
* show a spinner if it's running for too long
* become ready again once the action is complete (success or failure)

This would improve UX instantly while there are slowdowns, while still remaining useful in case we go from LLVM to monorepo with 3 million files.

With this being said, on internal channels I heard some operations may see a 10x performance improvement (🎉), but even with such a speedup, several minutes would still be many seconds.


### About LLVM-Project load times

While at it, I couldn't resist to profile it as well and it burns large amount of time in `git2::Repository::merge_base`.

<img width="1430" alt="Screenshot 2024-06-16 at 15 52 07" src="https://github.com/gitbutlerapp/gitbutler/assets/63622/081601f0-ab53-476a-8653-103351269fc7">

This is more of a note to self as I am collecting these issues at the moment. The UI is indeed unusable, as it times out and even when done quickly triggers the `list_virtual_branches` function again.

**Comment by krlvi** [comment]
Okay there is one more performance improvement which just landed with the latest patch release `0.12.5`. Let me know how it works for you

**Comment by PavelLaptev** [comment]
@krlvi cool! Can confirm that's pretty fast now ⚡

https://github.com/gitbutlerapp/gitbutler/assets/18498712/fc016582-a8d7-4af0-add6-ee4d73f2ea7b

**Comment by andrew-rosca** [comment]
I can also confirm the latest release is dramatically better!

**Comment by krlvi** [comment]
Keeping this issue open, but wanted to note that as of release 0.12.22 the performance on larger repos should be further improved

**Comment by andrew-rosca** [comment]
The last few releasees have worked great on the large(ish) repos I normally work with, with no noticeable impact on usability. There's still a slight lag when renaming branches, for example, but not enough to cause problems. From my perspective this issue is resolved.

If further improvements are possible and desirable, I suggest handling them as a separate story with measurable acceptance criteria.

**Comment by krlvi** [comment]
Thank you so much for getting back. Closing this now, feel free to open a new issue if you spot anything. Thanks

**Comment by moderniselife** [comment]
Version 0.13.2 - Takes about 7 second to detect changes to files and to detect that GitButler has made a commit.

I am happy to provide any information needed. 

Our repo has 64 branches ( they very much need to be deleted - so this could be an issue due to the branch count) 
The actual repo size is nearing the limit of having to switch to LFS.

**Comment by Byron** [comment]
Thanks for sharing! Does that mean GitButler felt faster before? Also, are there huge files either untracked or checked in to Git? How many files are in the index (`git ls-files | wc -l`)?

If it's a public repo it would certainly be great to have it for testing. I routinely test with the GitLab repository which currently shows very bad performance, which to me feels like a regression as well.

Currently, bad performance arises from a mix of…

* … a lot of tracked files
* … huge untracked or tracked files
* … remote is far ahead of virtual branches
* … many commits in virtual branches

What's not a problem though is many branches, this has been optimized and can handle tenth of thousands of them.

**Comment by moderniselife** [comment]
Hey!

Yup, the previous version were faster in terms of committing, but the version that was just released `0.13.4` improved speeds for loading the repo and recognising file changes but slowed down the time to commit and push.

Files in the index: 5205
Our node_modules folder is almost a gig, but we have that under gitignore, we do sync our firebase emulators to git to keep our devs in sync - this could be a big factor as it's export of the whole firestore emulator

Our repo isn't the largest but is definitely large.

My main/remote branch is our production branch so that is a good 200 commits behind the branch i am currently working on - not sure if that makes a difference.

My branch i am working on currently has 323 commits but it was an existing branch before I started using gitbutler.

![image](https://github.com/user-attachments/assets/94dbbecb-dcfe-4519-a021-cf8edbbb4ca2)

**Comment by Byron** [comment]
Thanks a lot for the heads-up! For now, I wouldn't know what else to ask, but just tried the latest released version on my GitLab repository and it's still unusable. While that's the case, this is what I can focus on and maybe that will also improve what you are encountering.

By looking at the sidebar it's clear that there are a lot of changes, and this combined with many commits might cause the slowdown.

5k files isn't that much, but maybe enough to make it struggle. It also does respect `.gitignore` so that *shouldn't* be an issue.

**Comment by moderniselife** [comment]
Alright awesome!

I am on macOS so i can get an instruments reading for you.

I'll get that to you soon!

**Comment by Byron** [comment]
Thanks for asking, and please feel free to open a new ones. The GitButler of today is quite different from what it was in this issue.
